### ISSUE TITLE
Vuejs 2.0 Migration

### ISSUE THREAD

**Issue by Humeira** [yuche/vue-strap#378]
I have started working on migrating **vue-strap** to vue 2.0, I'll be grateful if someone can create a new branch [v2](https://github.com/Humeira/vue-strap/tree/v2) here.

Some folks have show interest in migrating to 2.0 as well
https://github.com/yuche/vue-strap/issues/372
https://github.com/yuche/vue-strap/issues/374

**Comment by purepear** [comment]
I just found today this one for vue 2 and bootstrap 4
https://github.com/pi0/bootstrap-vue
I haven't tested it properly yet cause it's tricky to run it with vueify and without scss.. but i think if you use webpack it's good to go

**Comment by Humeira** [comment]
@purepear Yes it kind of have the same thing as **vue-strap**. But, **vue-strap** has more users I reckon it will be a really idea to start migrating this one, than moving to another completely different repo.

**Comment by Humeira** [comment]
@wffranco Hello is it possible for you to create a new branch **v2** on this repo?

**Comment by yuche** [comment]
@Humeira done.

**Comment by Humeira** [comment]
@yuche cool thanks!

**Comment by wffranco** [comment]
Well, i done a vue2 branch and started the migration, not finished yet.
https://github.com/wffranco/vue-strap/tree/vue2

**Comment by wffranco** [comment]
yep when i finish the migration i do the merge :smile:

**Comment by Humeira** [comment]
@wffranco you have already done most part? 🤔  i just started doing the same thing.

**Comment by psi-4ward** [comment]
so we should probably make it more transparent and list alle components in the first comment of this issue and track the progress here.

**Comment by wffranco** [comment]
I migrate all the components, but had a lot of errors in the documentation, the first one I don't undestrtand it...

I maybe will test the components apart of the project, to know if they work right.

If you can help me to detect what happen with the docs I appreciate your help.

**Comment by Humeira** [comment]
@wffranco where can i get your changes?

**Comment by druppy** [comment]
@wffranco I have cloned you repos, and it build nicely locally, but then i tried the [vue-migration-helper](https://github.com/vuejs/vue-migration-helper), and it came up with 166 errors.

Especially all the errors related to  [coerce](http://vuejs.org/guide/migration.html#coerce-Prop-Option-deprecated) properties and `twoWay`need some attention.
 
The Datepicker.vue has this [dispatch error](http://vuejs.org/guide/migration.html#dispatch-and-broadcast-deprecated).

If we could find a common solution for these errors, then it is just a question of fixing the Components one at a time (make a list and spit it in 3 ?).

**Comment by wffranco** [comment]
Sorry I'm really busy with 2 projects, so I have no time to continue the migration right now.
@druppy I run the migration helper 3 days ago in src and fix all the errors, let me check if I forgot to push the last changes.

**Comment by ropeladder** [comment]
Can confirm that the `src` directory of the `vue2` branch on wffranco's fork returns no errors. I'm getting a build error though because for some reason `ButtonGroup.vue` got renamed to `buttonGroup.vue`; otherwise it seems to build fine.

**Comment by purepear** [comment]
@ropeladder haven't tried it yet but might be a linux/mac/windows - uppercase/lowercase filesystem thing

**Comment by wffranco** [comment]
I think the best option is testing the components individually, maybe in a clean window...
But like I said I have no time right now.

**Comment by druppy** [comment]
@wffranco Nice work !, please let me know where to get these changes, or when you have time to commit it onto the fork. I have cloned your fork, so I can test it from there.

@ropeladder how did you test that ? I still get 166 errors from the `vue-migration-helper` tool :-)

**Comment by ropeladder** [comment]
@druppy Just like this:
```
git clone https://github.com/wffranco/vue-strap.git
cd vue-strap
git checkout vue2
cd src
vue-migration-helper
```

**Comment by wffranco** [comment]
@ropeladder you right I renamed the buttongroup, now match with the standard in the project.
But the real problem is that I comment almost all the documentation files and still not working... so I think that need to be tested one by one.

**Comment by druppy** [comment]
@ropeladder Sorry, I still need to work at my black belt as git ninja :-) git will not checkout is` dist` has some build js files ... sight. 

So, time to test components ... thanks !

**Comment by ropeladder** [comment]
I'm trying out dropdowns and they appear to work correctly, but Vue gives a warning that they should not be mutating the `value` property. I don't really understand well enough how the component is built to know what the correct way to refactor would be.

**Comment by wffranco** [comment]
well, I needed select component so I fix it and fixed the helpers for supporting of laravel 5.3 (give some errors because this uses buble-loader to convert ECS2015 and have some limitations).

nobody else have working over my branch? if somebody want to help just tell me or make a push request if you're done something.

Here the branch again if somebody need it: https://github.com/wffranco/vue-strap/tree/vue2

**Comment by wffranco** [comment]
acording to vue2, twoway binding is not supported, except for v-model.
but I think on a simple feature to make a twoway binding for any propertie, compatible with vue 1 and 2, but not tested yet.

**Comment by druppy** [comment]
@wffranco I have just tried to fix a few build docs errors, but I have no access to the branch :-)

ERROR: Permission to wffranco/vue-strap.git denied to druppy.

**Comment by wffranco** [comment]
make a pull request?
anyway, let me see if I can add you :stuck_out_tongue:

**Comment by druppy** [comment]
Yes, thanks ... tried some small changes, time for some more work, now that it is not in vain :-)

Any reason for `dist` and `build` content to be added to the repos, git complains about it after a re-build.

**Comment by wffranco** [comment]
- fixed Input
- fixed select
- formGroup will be renamed to formValidator, the formGroup component will have another purpose
- reordering components for a better control of the fixes

By for now I only will fix the components I need, have a lot of work, but need some components in vue 2.
Next to fix:
- tabs
- tab

**Comment by webnoob** [comment]
@wffranco Is there a way to see everything that *has* been ported to vue2 and what is left outstanding?

I'd like to start using this in my project but need to know what's available (and I might be able to help with the migration for the bits I need personally).

Also, with regards to using the Vue2 version, how do we go about that, just `npm install vue-sharp` as normal or is there a special version we should be installing?

**Comment by Humeira** [comment]
@webnoob It's still a work in progress. You can find the changes on @wffranco branch `vue2`. 
If you want to see `vue-strap` with vue.js 2.x, just switch to that branch and do a `npm install`. Everything already in the `package.json`.

**Comment by druppy** [comment]
I have tried to fix `npm run docs` so that it again works and compiles against `vue2`, this make testing much more simple :) Hope others may find this useful too ...

**Comment by webnoob** [comment]
@druppy Thanks! I too had noticed it wasn't working and thought it might be useful to have it fixed.

**Comment by wffranco** [comment]
I finally made the docs work in vue 2.
But I detected that selectDocs crash my chrome. So I commented the most of the documentations.
When you enable something, try to do it one by one.

**Comment by wffranco** [comment]
Fixed:
- dropdown
- tab
- tabGroup
- tabs

Fixed, but have to check them again:
- input
- select
Input return some errors in the docs, and selectDocs crash so must be something in the select component.

**Comment by druppy** [comment]
@wffranco nice with some progress :-) We ended up changing in the same part of the code, lets see if we can avoid that, in the name of progress. I had the whole docs page rendering with all exiting components, but it is ok to just enable them one at a time.

I will try to focus on the `Modal` then, as you are working on the other important parts.

**Comment by webnoob** [comment]
I'd be eternally grateful if someone could work on the Datepicker :)

**Comment by druppy** [comment]
@webnoob Noted :-)

**Comment by wffranco** [comment]
@druppy if you handle the datepicker I have an idea for a long time but never had time to do it, and is convert the datepicker in a wrapper, something closer to the tooltip / popover, then we can use the datepicker with an input or a button (in some mobile phones the native datepicker looks very good, so I think that in a button you can chose what datepicker to use).

**Comment by webnoob** [comment]
Just an FYI, I'm occasionally getting this on v-select:
`
vue-strap.min.js:2 Uncaught (in promise) TypeError: $(...).offBlur is not a function(…)beforeDestroy @ vue-strap.min.js:2callHook @ vue.common.js?e881:1795Vue.$destroy @ vue.common.js?e881:1759destroy$1 @ vue.common.js?e881:1968invokeDestroyHook @ vue.common.js?e881:3883invokeDestroyHook @ vue.common.js?e881:3887invokeDestroyHook @ vue.common.js?e881:3891invokeDestroyHook @ vue.common.js?e881:3891invokeDestroyHook @ vue.common.js?e881:3891invokeDestroyHook @ vue.common.js?e881:3891invokeDestroyHook @ vue.common.js?e881:3891invokeDestroyHook @ vue.common.js?e881:3891invokeDestroyHook @ vue.common.js?e881:3891invokeDestroyHook @ vue.common.js?e881:3887invokeDestroyHook @ vue.common.js?e881:3887removeVnodes @ vue.common.js?e881:3902updateChildren @ vue.common.js?e881:4009patchVnode @ vue.common.js?e881:4042updateChildren @ vue.common.js?e881:3960patchVnode @ vue.common.js?e881:4042patch @ vue.common.js?e881:4151Vue._update @ vue.common.js?e881:1689(anonymous function) @ vue.common.js?e881:1662get @ vue.common.js?e881:729run @ vue.common.js?e881:798flushSchedulerQueue @ vue.common.js?e881:616nextTickHandler @ vue.common.js?e881:396`

I'm using hot reload so I don't know if that's the issue but it breaks the page from then on.

**Comment by druppy** [comment]
@wffranco I have worked a bit on the datepicker (included in docs again), and it seems to work, but select are strange and I still have a some mutate warnings, that I need to understand more.

**Comment by madr** [comment]
FWIW, I have tested the dist files in the vue2 branch in @wffranco s fork, and I noticed that the input field of the typeahead component silently ignores all keydowns and keyups: you type, but no chars appear in the input field.

I have just started with Vue and have no clue, sadly. I could take a look though if someone might give me a hint.

**Comment by wffranco** [comment]
@webnoob I don't compiled the vuestrap by for now because is not ready. I don't know if @druppy do it, but if he don't, the vue-strap.min.js will not work.

**Comment by druppy** [comment]
@wffranco I don't compile it either  I use the `npm run docs` until all works ... hopefully soon.

**Comment by webnoob** [comment]
@wffranco @druppy I cloned the source and did "npm run build" myself so the vue-strap.min.js file should be Ok shouldn't it (assuming I stick to only those components that have been migrated). I'm not getting any other errors using those components and I would have expected a lot if I was using components that have not yet been migrated.

**Comment by druppy** [comment]
@webnoob Thanks, so nice to hear, @wffranco have made a lot of this possible

**Comment by IanGrainger** [comment]
Ah - this might partly answer a question I just put on SO.com here: http://stackoverflow.com/questions/40266688/typeahead-with-vue-js-2-0 ;)

**Comment by webnoob** [comment]
@druppy @wffranco There is definitely an issue with the select component. I've updated my source and re-built. The offending line is line 242 in Select.vue `$(this.$refs.select).offBlur()`. I've commented the line out for now in my version whilst I'm developing but noting here so it can be tracked down later (as I know it's a component that has already been migrated)

**Comment by druppy** [comment]
@wffranco I have added a Modal that works ... once. I have changed it to let the open state be handled by the parent, as you can see in the docs. If you could give me an idea as to why it will not open second time I would be happy to fix it.

When all this is done, there may be an idea in remove the pseudo jQuery layer and make a prober Vue2 code base :-)

**Comment by wffranco** [comment]
@druppy move to the branch v2
https://github.com/wffranco/vue-strap/tree/v2

And check the input, I was working over it and you too... I mix both but I think your blur/focus events are not necessary, I was fixed that before

**Comment by druppy** [comment]
@wffranco ok, I moved to v2, and will look at input and the blur/focus events.

The Accordion.vue works for me, but the css effect are strange, any idear :-) 

Select is also strange, I will try to look at its auto close function.

I try to remove as much dependency on NodeList.js as possible, and a saw @wffranco has tried to use only one single version. The more we can use Vue2 as it is intended the better :-)

Could we get a vue-strap2 gitter or something ?

**Comment by wffranco** [comment]
i make a fast change in the select. removed onblur nodelist event, and used a new directive.
@druppy test it if you can. going late to work 🤓

**Comment by wffranco** [comment]
🦀 have some errors, this branch was created over the yuche v2 instead of vue2 like I think and now have conflict... i will add it later

**Comment by wffranco** [comment]
Fixed.

**Comment by druppy** [comment]
@wffranco Select works really nice now, it hides selections as expected !

**Comment by webnoob** [comment]
So, are you guys working on v2 now instead of vue2 branch?

**Comment by druppy** [comment]
@webnoob yeps @wffranco's fork and the v2 branch. As I understand it, this was to harmonize the branch name with the original repos, and make PR more straightforward.

**Comment by webnoob** [comment]
Ah, ok. Thanks. I'll pull it later and see how the new selects work :)

**Comment by wffranco** [comment]
Affix now work fine.

@druppy , the most of the things that need changes is for twoway binding or modify properties inside the component.
I think that the properties must be closer to vue1 version, so try to do that:
- create a new variable in the component data (e.g. for property `value`, create variable `val` in data)
- add a watcher to the propertie that update the variable.
- add a watcher to the variable that emit an event with the new value
- in the specific case of value (for v-model binding) remember to emit `input` event
Example of that on input/select components.

**Comment by wffranco** [comment]
Typeahead fixed.
- async typeahead need some adjustment for lazy loads, but the old functionality is done.

**Comment by druppy** [comment]
@wffranco Affix works here too, nice. 

Panel isOpen fixed, alert doc duration fix, and very basic typescript support.

**Comment by druppy** [comment]
Checkbox + docs works, carousel has no more warning, not sure how the indicators should have worked :-) 

We have changed `checked` prob to Boolean but the emitted event is value or null, is this what we want ?

**Comment by wffranco** [comment]
I saw that you added `typings` to the project....

What is it used for?

**Comment by druppy** [comment]
Typings are new to TS 2 and makes it possible to include typescript definitions in npm for seamless integration, [see this](http://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html).

vue, vue-router and vue-class-component all have this, and combined with npm @typings we can now enjoy typescript in a pure npm setup :-)

**Comment by wffranco** [comment]
merged v2 branch. set the version to 2.0.0-pre.1
Still need some fixes but working the most. I stablished vue2 as master in my fork.

**Comment by druppy** [comment]
Ok, looks really good, thanks for all the hard work. 

Do you have any idea as to when `vue-strap` will pull in you changes and officially support Vue2 ?

**Comment by webnoob** [comment]
@wffranco Really appreciate all the work you've put in. Thanks.

**Comment by FerchoCarcho** [comment]
if Npm it will it bring the v.2?

**Comment by webnoob** [comment]
@druppy With regards to your comment about making this being merged into the main vue-strap brance and support vue2, I think we're still a little ways off that. There are still a number of warnings being raised when using the components (i.e Datepicker warns about mutating props directly) and also raises an error about window.eventlistener.

There are some other issues as well but thought I'd note here so people don't think it's production ready just yet.

**Comment by druppy** [comment]
@webnoob I am aware of this, but I guess @wffranco are cleaning up details now, and prepares for the final PR :-) I am just using the fork as a local clone (v2 branch), until a release happens :-)

**Comment by webnoob** [comment]
@druppy Yeah understandable, given that @FerchoCarcho already asked about using this I thought it wise to make sure it's clear it's not production ready just yet :) Close though! :P

**Comment by wffranco** [comment]
Yuche has been inactive for several months, so at the moment I prefer to work on my branch and update eventually over the yuche/v2 branch.
@druppy like I said the last time, I'm working over my master branch, so if you're changing something please try to do it there or notify me to merge 😃

**Comment by druppy** [comment]
@wffranco I will switch from `v2` to `master` branch in your fork, I was just confused ... again :-) Anyway, it begins to look really good now, except for small details.

**Comment by Humeira** [comment]
@wffranco he replied on this thread like a month ago, maybe another ping @yuche !

Thanks for the work!! 🎉

**Comment by mediaessenz** [comment]
@wffranco could you please active issue tracker inside your fork?
This would make collaboration easier 😄

**Comment by wffranco** [comment]
done 👍

**Comment by otabekgb** [comment]
This repository seems to have stuck. Community should move to wffranco. How can we inform the whole community?

**Comment by purepear** [comment]
@druppy @wffranco about Datepicker 
https://github.com/wffranco/vue-strap/issues/8 
https://gist.github.com/purepear/df5fb8a6cc862071bb07ca5d059a0ade

**Comment by neilsmind** [comment]
@wffranco - Really appreciate your effort to move things forward to v2. Noticed that you [merged v2](https://github.com/yuche/vue-strap/commit/df4668271342d430333c3b4b69630bafc0de7c47) as a [branch](https://github.com/yuche/vue-strap/tree/v2) into this repo in March.

Given that your project was merged back here, no further commits have been made there since then and you're a collaborator on this repository, does this repo remain core for vue-strap and, more importantly, v2?  

We're using vue-strap within a project and I'm happy to contribute / collaborate...just trying to figure out where. :-)
